### ISSUE TITLE
Episodic Memory & Learning

### ISSUE THREAD

**Issue by c-daly** [c-daly/logos#414]
## Summary

Enable the agent to learn from execution history and improve plan quality over time through episodic memory and probabilistic validation.

**Spec Reference:** [`docs/architecture/PHASE3_SPEC.md`](https://github.com/c-daly/logos/blob/main/docs/architecture/PHASE3_SPEC.md)

---

## Goals

Transform LOGOS from a stateless planner into a learning system that:
- Records execution episodes (plan → actions → outcomes)
- Queries similar past experiences before generating new plans
- Improves plan quality measurably over repeated similar goals
- Adjusts capability confidence based on historical success rates
- Layers probabilistic validation on top of SHACL constraints

---

### Episodic Storage
- [ ] HCG schema for Episode nodes
- [ ] CRUD operations for episodes
- [ ] Episode indexing by goal, capabilities, context
- [ ] Relationship to PersonaEntry reflections

### Episode Structure

```python
Episode {
  id: UUID
  timestamp: datetime
  goal: Goal
  plan_id: UUID
  actions: List[{
    capability: str
    params: dict
    result: str  # success/failure
    confidence: float
  }]
  outcome: {
    success: bool
    failure_reason: Optional[str]
    duration_ms: int
  }
  context: {
    user: Optional[str]
    environment: str
    constraints: List[str]
  }
  reflections: List[UUID]  # Related PersonaEntry IDs
  learned: {
    patterns: List[str]
    adjustments: List[str]
    recommendations: List[str]
  }
}
```

### Learning Integration
- [ ] Planner queries relevant episodes before generating new plans
- [ ] Capability confidence scores adjust based on historical outcomes
- [ ] Context-sensitive learning: Different strategies for different domains/users
- [ ] Pattern detection across multiple episodes

### Probabilistic Validation (Level 2)
- [ ] Augment SHACL with uncertainty-based validation
- [ ] Use learned patterns to estimate plan success probability
- [ ] Flag uncertain plans for human review or alternative generation
- [ ] Track validation accuracy over time

---

## Learning Flow

```
1. Plan execution completes (success or failure)
       │
       ▼
2. Create Episode record with full trace
       │
       ▼
3. Link to related PersonaEntry reflections
       │
       ▼
4. Index by goal type, capabilities, context
       │
       ▼
5. Future plans query similar episodes
       │
       ▼
6. Adjust confidence, strategy, capability selection
```

---

## Success Criteria

- [ ] Plan quality measurably improves over repeated similar goals
- [ ] Probabilistic validation catches issues SHACL misses
- [ ] Capability selection adapts based on historical success rates
- [ ] Agent demonstrates learning **without explicit retraining**
- [ ] Episode queries complete in <100ms for typical workloads

---

## Effort Estimate

**3-4 weeks**

---

## Dependencies

- **P3-M1** (Hierarchical Memory) - Episode storage in long-term tier
- **P3-M2** (Event-driven Reflection) - Reflections linked to episodes
- **P3-M3** (Selective Diary) - Completed for consistency

---

### Episode Indexing Strategy

Index episodes for efficient retrieval:
- **By goal type**: "pick_and_place", "navigation", "query_answer"
- **By capability set**: Hash of capabilities used
- **By context**: User, environment constraints
- **By outcome**: Success/failure for pattern analysis

### Similarity Search for Episodes

When planning:
```python
def find_similar_episodes(goal: Goal, context: Context) -> List[Episode]:
    # 1. Match on goal type (exact)
    # 2. Embed goal description for semantic similarity
    # 3. Filter by context overlap
    # 4. Rank by recency and relevance
    # 5. Return top-k for planner consideration
```

### Probabilistic Validation

Complement SHACL structural validation with learned probability:
```python
def validate_plan(plan: Plan) -> ValidationResult:
    shacl_result = run_shacl_validation(plan)
    
    similar_episodes = find_similar_episodes(plan.goal)
    success_rate = calculate_success_rate(similar_episodes, plan)
    
    return ValidationResult(
        shacl_valid=shacl_result.valid,
        estimated_success=success_rate,
        confidence_interval=(lower, upper),
        similar_failures=extract_failure_patterns(similar_episodes),
        recommendations=generate_recommendations(similar_episodes)
    )
```

---

## Metrics & Monitoring

- Episode count by goal type
- Success rate trends over time
- Plan improvement ratio (2nd attempt vs 1st)
- Capability confidence distribution
- Probabilistic validation accuracy

---

## Future Extensions (Phase 4+)

- Memory compression: Summarize old episodes
- Multi-agent knowledge transfer
- Meta-learning: Learn learning rate itself
- Curiosity-driven exploration to fill knowledge gaps

---

## References

- [PHASE3_SPEC.md](https://github.com/c-daly/logos/blob/main/docs/architecture/PHASE3_SPEC.md) - Full specification
- [HCGPlanner](https://github.com/c-daly/logos/blob/main/logos_hcg/planner.py) - Current planner implementation

---

**Project:** [LOGOS](https://github.com/users/c-daly/projects/2)

**Comment by c-daly** [comment]
## Status Update (2026-02-28) — Triage

**Status: Not started.** Renamed from "P3-M4" — phase language deprecated.

No episodic memory, episode replay, or pattern extraction from history exists. The closest thing is an experimental PoC in sophia (`sophia/src/sophia/experiments/emotional_drift.py`) exploring Hebbian-style embedding drift, but it's a research prototype not integrated into any production path.

This is the final milestone in the memory & learning sequence, depending on #411, #412, #413.
